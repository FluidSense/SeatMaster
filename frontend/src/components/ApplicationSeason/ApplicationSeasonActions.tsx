import { AnyAction, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { GET_APPLICATION_SEASON, SET_APPLICATION_SEASON } from './Strings';

export const fetchApplicationSeasonData = ():
  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
    fetch(GET_APPLICATION_SEASON)
      .then(response => response.json())
      .then(result => dispatch({
        payload: result,
        type: SET_APPLICATION_SEASON,
      }));
  };
