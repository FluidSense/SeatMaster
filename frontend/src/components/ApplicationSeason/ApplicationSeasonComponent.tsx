import EtikettBase from 'nav-frontend-etiketter';
import React, { Fragment } from 'react';
import { ETIKETT_YELLOW, SEASON_CLOSED, SEASON_END, SEASON_START } from './Strings';

interface IApplicationSeasonState {
  seasonStart: Date;
  seasonEnd: Date;
}

type validPreText = typeof SEASON_START | typeof SEASON_END;

const formatSeasonString = (preText: validPreText, time: Date) => {
  const day = time.getDate();
  const month = time.getMonth() + 1;
  const year = time.getFullYear();
  return `${preText} ${day}/${month}/${year}`;
};

const NavEtikett = (seasonStatus: string) => {
  return (
    <EtikettBase
      type={ETIKETT_YELLOW}
    >
      {seasonStatus}
    </EtikettBase>
  );
};

const ApplicationSeasonComponent: React.FunctionComponent<IApplicationSeasonState> = (props) => {
  const { seasonEnd, seasonStart } = props;
  const currentTime = new Date();
  const seasonEndComponent = currentTime < seasonEnd
    ? NavEtikett(formatSeasonString(SEASON_END, seasonEnd)) : NavEtikett(SEASON_CLOSED);
  const seasonStartComponent = currentTime < seasonStart
    ? NavEtikett(formatSeasonString(SEASON_START, seasonStart)) : null;
  return(
    <Fragment>
      {seasonStartComponent}
      {seasonEndComponent}
    </Fragment>
  );
};

export default ApplicationSeasonComponent;
