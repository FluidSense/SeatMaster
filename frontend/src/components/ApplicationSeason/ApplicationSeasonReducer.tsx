import { SET_APPLICATION_SEASON } from './Strings';

export interface IApplicationSeasonState {
  applicationSeason?: {
    applicationPeriodEnd: Date;
    applicationPeriodStart: Date;
    start: Date;
    end: Date;
  };
}

export const initialState = {
  applicationSeason: undefined,
};

export const applicationSeasonReducer = (
  state: IApplicationSeasonState = initialState,
  action: any,
  ): IApplicationSeasonState => {
  const { type, payload } = action;
  switch (type) {
    case SET_APPLICATION_SEASON:
      const {
        applicationPeriodEnd,
        applicationPeriodStart,
        start,
        end,
      } = payload;
      return {
        ...state,
        applicationSeason: {
          applicationPeriodEnd: new Date(applicationPeriodEnd),
          applicationPeriodStart: new Date(applicationPeriodStart),
          end: new Date(end),
          start: new Date(start),
        },
      };
    default:
      return state;
  }
};
