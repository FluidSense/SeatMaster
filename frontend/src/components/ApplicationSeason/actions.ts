import { AnyAction, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { getSeason, getSeasonFromId } from '../../API/calls';
import isEmpty from '../../utils/objectIsEmpty';
import { FETCHED_APPLICATION_SEASON, SET_APPLICATION_SEASON } from './constants';
import { IApplicationSeason } from './reducer';

const setApplicationSeason = (payload: IApplicationSeason) => ({
  payload,
  type: SET_APPLICATION_SEASON,
});

const fetchedApplicationSeason = (payload: IApplicationSeason) => ({
  payload,
  type: FETCHED_APPLICATION_SEASON,
});

export const fetchApplicationSeasonData = ():
  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
    const result = await getSeason();
    if (result && !isEmpty(result)) dispatch(setApplicationSeason(result));
  };

export const fetchSeasonById = (id: number):
ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
  const result = await getSeasonFromId(id);
  if (result && !isEmpty(result)) dispatch(fetchedApplicationSeason(result));
};
