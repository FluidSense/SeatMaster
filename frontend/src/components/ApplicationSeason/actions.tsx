import { AnyAction, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { SEASON_API_URL, SET_APPLICATION_SEASON } from './constants';

export const fetchApplicationSeasonData = ():
  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
    fetch(`${SEASON_API_URL}getSeason`)
      .then(response => response.json())
      .then(result => dispatch({
        payload: result,
        type: SET_APPLICATION_SEASON,
      }));
  };
