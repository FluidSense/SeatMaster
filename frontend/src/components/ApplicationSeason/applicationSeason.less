#season-info {
  display: flex;
  flex-direction: column;
}
