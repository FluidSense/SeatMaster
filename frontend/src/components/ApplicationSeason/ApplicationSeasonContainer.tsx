import { connect } from 'react-redux';
import ApplicationSeasonComponent from './ApplicationSeasonComponent';

// FIXME state should not be any, WIP.
const mapStateToProps = (state: any) => ({
  applicationSeason: state.home.applicationSeason,
});

const ApplicationSeasonContainer = connect(
  mapStateToProps,
  null,
)(ApplicationSeasonComponent);

export default ApplicationSeasonContainer;
