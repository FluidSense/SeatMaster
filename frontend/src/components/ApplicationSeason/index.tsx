import moment from 'moment';
import React from 'react';
import { connect } from 'react-redux';
import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { fetchApplicationSeasonData } from './actions';
import './applicationSeason.css';
import Presentational from './Presentational';
import { IApplicationSeason } from './reducer';

interface IStateProps {
  applicationSeason: IApplicationSeason;
}

interface IDispatchProps {
  fetchSeason: () => ThunkAction<Promise<AnyAction>, {}, {}, AnyAction>;
}

type Props = IStateProps & IDispatchProps;

// tslint:disable-next-line:class-name
class _Container extends React.Component<Props, {}> {
  public componentDidMount = () => {
    const { fetchSeason } = this.props;
    fetchSeason();
  }

  public render = () => {
    const { applicationSeason } = this.props;
    return (
      <Presentational applicationSeason={applicationSeason} currentDate={moment()} />
    );
  }
}

const mapStateToProps = (state: IStateProps) => ({
  applicationSeason: state.applicationSeason,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, any>) => ({
  fetchSeason: () => dispatch(fetchApplicationSeasonData()),
});

const ApplicationSeason = connect(
  mapStateToProps,
  mapDispatchToProps,
)(_Container);

export default ApplicationSeason;
