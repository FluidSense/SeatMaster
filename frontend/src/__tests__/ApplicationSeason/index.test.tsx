import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import moment from 'moment';
import * as React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ApplicationSeason from '../../components/ApplicationSeason';

describe('ApplicationSeasonContainer', () => {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore(
    {
      applicationSeason: {
        currentSeason: {
          applicationPeriodEnd: moment.utc('20190202'),
          applicationPeriodStart: moment.utc('20190202'),
          end: moment.utc('20190202'),
          start: moment.utc('20190202'),
        },
      },
    });

  // Overwrite Date.now being used by moment() to always return same value
  Date.now = jest.fn(() => new Date(Date.UTC(2017, 0, 1)).valueOf());
  it('renders correctly', () => {
    const wrapper = mount(
      <Provider store={store} >
        <ApplicationSeason />
      </Provider>,
    );
    const container = wrapper.find(ApplicationSeason);
    expect(container.length).toBeTruthy();
  });

  it('matches snapshot', () => {
    const wrapper = mount(
      <Provider store={store} >
        <ApplicationSeason />
      </Provider>,
    );
    const container = wrapper.find(ApplicationSeason);
    expect(toJson(container)).toMatchSnapshot();
  });
});
