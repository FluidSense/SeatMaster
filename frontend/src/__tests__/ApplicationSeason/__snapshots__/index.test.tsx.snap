// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationSeasonContainer renders correctly 1`] = `
<Connect(_Container)>
  <_Container
    applicationSeason={
      Object {
        "applicationPeriodEnd": "2019-02-02T00:00:00.000Z",
        "applicationPeriodStart": "2019-02-02T00:00:00.000Z",
        "end": "2019-02-02T00:00:00.000Z",
        "start": "2019-02-02T00:00:00.000Z",
      }
    }
    fetchSeason={[Function]}
  >
    <Presentational
      applicationSeason={
        Object {
          "applicationPeriodEnd": "2019-02-02T00:00:00.000Z",
          "applicationPeriodStart": "2019-02-02T00:00:00.000Z",
          "end": "2019-02-02T00:00:00.000Z",
          "start": "2019-02-02T00:00:00.000Z",
        }
      }
      currentDate={"2017-01-01T00:00:00.000Z"}
    >
      <EtikettBase
        type="info"
        typo="normaltekst"
      >
        <div
          className="etikett etikett--info"
        >
          <TypografiBase
            tag="span"
            type="normaltekst"
          >
            <span
              className="typo-normal"
            >
              Open for applications  - 2/2/2019
            </span>
          </TypografiBase>
        </div>
      </EtikettBase>
      <EtikettBase
        type="info"
        typo="normaltekst"
      >
        <div
          className="etikett etikett--info"
        >
          <TypografiBase
            tag="span"
            type="normaltekst"
          >
            <span
              className="typo-normal"
            >
              Closing for applications  - 2/2/2019
            </span>
          </TypografiBase>
        </div>
      </EtikettBase>
    </Presentational>
  </_Container>
</Connect(_Container)>
`;
